src/Data/Aeson/Types/FromJSON.hs

```diff
@@ -1348,10 +1360,17 @@ instance ( RecordFromJSON' arity a
 
 instance {-# OVERLAPPABLE #-} (Selector s, GFromJSON arity a) =>
          RecordFromJSON' arity (S1 s a) where
-    recordParseJSON' (cname :* tname :* opts :* fargs) obj = do
-        fv <- contextCons cname tname (obj .: label)
-        M1 <$> gParseJSON opts fargs fv <?> Key label
+    recordParseJSON' (cname :* tname :* opts :* fargs) obj
+      | requireOptionalFields opts = requireField
+      | label `KM.member` obj = requireField
+      | Nothing <- maybeDef = requireField
+      | Just def <- maybeDef = pure def
```

Comment:
This feels a little obtuse to parse - logically, it's:

```haskell
if requireOptionalFields opts || label `KM.member` obj
  then requireField
  else case maybeDef of
    Nothing -> requireField
    Just def -> pure def
```

But, what we're really varying over here is how the field access goes on about - `contextCons cname tname (obj .: label)`. All `contextCons` does is put some error message info on the underlying parser, so we could fold that logic into the `Parser` there.

The function `.:?` has the behavior we want, sort of:

```haskell
-- | Retrieve the value associated with the given key of an 'Object'. The
-- result is 'Nothing' if the key is not present or if its value is 'Null',
-- or 'empty' if the value cannot be converted to the desired type.
--
-- This accessor is most useful if the key and value can be absent
-- from an object without affecting its validity.  If the key and
-- value are mandatory, use '.:' instead.
(.:?) :: (FromJSON a) => Object -> Key -> Parser (Maybe a)
(.:?) = explicitParseFieldMaybe parseJSON
```

So, this means we can write it out as:

```haskell
conditionalParseField :: FromJSON a => Options -> Object -> Key -> Parser a
conditionalParseField opts obj key
  | requireOptionalFields opts =
      obj .: key
  | otherwise =
       case maybeDef of
         Nothing -> obj .: key
          Just def -> 
            fromMaybe def <$> obj .:? key
```